### PR TITLE
English

### PR DESCRIPTION
**Pull request by YCGreen**

### CODE REVIEW

File: src/main/java/green/dictionary/EnglishDictionary.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import com.opencsv.CSVReader;
+import com.opencsv.exceptions.CsvValidationException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+public class EnglishDictionary {
+    private static final HashMap<String, List<String[]>> definitions = new HashMap<>();
```
Comment (andrewoid) [inline src/main/java/green/dictionary/EnglishDictionary.java]: Why is this static?

File: src/test/java/green/dictionary/EnglishDictionaryTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        // given
+        EnglishDictionary dictionary = new EnglishDictionary();
+
+        // when
+        List<String[]> definitions = dictionary.getDefinition("Aband");
+
+        // then
+        assertEquals("Aband, v. t., To abandon."
+                + "Aband, v. t., To banish; to expel.",
+                dictionary.toString(definitions));
```
Comment (andrewoid) [inline src/test/java/green/dictionary/EnglishDictionaryTest.java]: Or you can just assertEquals each item in the array...

Comment (YCGreen) [reply src/test/java/green/dictionary/EnglishDictionaryTest.java]: Like per word?

File: src/main/java/green/dictionary/EnglishDictionary.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                wordDefs.add(nextDef);
+                definitions.put(word, wordDefs);
+            }
+        } catch (IOException | CsvValidationException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    public List<String[]> getDefinition(String word) {
```
Comment (andrewoid) [inline src/main/java/green/dictionary/EnglishDictionary.java]: getDefintion should only return the definition, not the whole line.

File: src/main/java/green/dictionary/Main.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    public static void main(String[] args) {
+        EnglishDictionary dic = new EnglishDictionary();
+
+        List<String[]> text = dic.getDefinition("Aard-vark");
+
+        System.out.println(dic.toString(text));
+
+
+
+    }
```
Comment (andrewoid) [inline src/main/java/green/dictionary/Main.java]: You have a test for this. 
Get out of the habit of writing mains that should just be tests

File: src/main/java/green/dictionary/EnglishDictionary.java [hunk 5]
```diff
@@ -12,9 +13,13 @@
 public class EnglishDictionary {
     private final HashMap<String, List<String[]>> definitions = new HashMap<>();
 
-    public EnglishDictionary()  {
-        try (CSVReader csvReader = new CSVReader(new FileReader(
-                "englishDictionary.csv"))) {
+    public EnglishDictionary() {
+        InputStream in = EnglishDictionary.class.getResourceAsStream(
+                "englishDictionary.csv");
```
Comment (andrewoid) [inline src/main/java/green/dictionary/EnglishDictionary.java]: try "/englishDictionary.csv"

File: src/main/java/green/dictionary/EnglishDictionary.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            return new String[] {"Abnormal format of definition, unable to extract"};
+        }
+        return Arrays.copyOfRange(line, 2, line.length);
+    }
+
+    public List<String[]> getDefinition(String word) {
+        return definitions.get(word);
+    }
+
+    public String toString(List<String[]> text) {
```
Comment (andrewoid) [inline src/main/java/green/dictionary/EnglishDictionary.java]: What does this method do?

File: src/test/java/green/dictionary/EnglishDictionaryTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    public void getDefinitionNonexistent() {
+        // given
+        EnglishDictionary dictionary = new EnglishDictionary();
+
+        // when
+        List<String[]> definitions = dictionary.getDefinition("blahblahblah");
+
+        // then
+        assertEquals(null, dictionary.toString(definitions));
```
Comment (andrewoid) [inline src/test/java/green/dictionary/EnglishDictionaryTest.java]: assertNull

File: src/test/java/green/dictionary/EnglishDictionaryTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    public void getDefinitionMultiple() {
+        // given
+        EnglishDictionary dictionary = new EnglishDictionary();
+
+        // when
+        List<String[]> definitions = dictionary.getDefinition("Aband");
+
+        // then
+        assertEquals("To abandon."
```
Comment (andrewoid) [inline src/test/java/green/dictionary/EnglishDictionaryTest.java]: Assert the individual elements of the array.

File: src/main/java/green/dictionary/EnglishDictionary.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    private String[] extractDefFromLine(String[] line) {
+        if (line.length < 3) { //default definition is 3rd element as per this csv
+            return new String[] {"Abnormal format of definition, unable to extract"};
+        }
+        return Arrays.copyOfRange(line, 2, line.length);
+    }
+
+    public List<String[]> getDefinition(String word) {
```
Comment (andrewoid) [inline src/main/java/green/dictionary/EnglishDictionary.java]: This method is supposed to return List<String>

File: src/test/java/green/dictionary/EnglishDictionaryTest.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Test
+    public void getDefinitionMultiple() {
+        // given
+        EnglishDictionary dictionary = new EnglishDictionary();
+
+        // when
+        List<String> definitions = dictionary.getDefinition("Aband");
+
+        // then
```
Comment (andrewoid) [inline src/test/java/green/dictionary/EnglishDictionaryTest.java:48]: toString() here is unnecessary. It's already a String

File: src/test/java/green/dictionary/EnglishDictionaryTest.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        EnglishDictionary dictionary = new EnglishDictionary();
+
+        // when
+        List<String> definitions = dictionary.getDefinition("Aard-vark");
+
+        // then
+        assertEquals("An edentate mammal, of the genus Orycteropus, "
+                + "somewhat resembling a pig, common in some parts of Southern Africa. "
+                + "It burrows in the ground, and feeds entirely on ants, "
+                + "which it catches with its long, slimy tongue.", definitions.get(0).toString());
```
Comment (andrewoid) [inline src/test/java/green/dictionary/EnglishDictionaryTest.java]: Same as below

File: src/main/java/green/dictionary/Main.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package green.dictionary;
+
+import com.opencsv.exceptions.CsvException;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class Main {
```
Comment (andrewoid) [inline src/main/java/green/dictionary/Main.java:3]: File is not needed

File: src/main/java/green/dictionary/EnglishDictionary.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private String extractDefFromLine(String[] line) {
+        if (line.length < 3) { //default definition is 3rd element as per this csv
+            return "Abnormal format of definition, unable to extract";
+        }
+        return String.join(" ", Arrays.copyOfRange(line, 2, line.length));
```
Comment (andrewoid) [inline src/main/java/green/dictionary/EnglishDictionary.java]: Why are you doing this now?

**Review by andrewoid — APPROVED** [review]

**Pull request merged**